### ISSUE TITLE
3D surf/pcolor plots are too small

### ISSUE THREAD

**Issue by cbm755** [Calysto/octave_kernel#41]
This makes a very small image:

```
x1 = linspace(-1,1, 11);
[x,y] = meshgrid(x1,x1);
h = pcolor(x,y,sin(x)+y);
```

This one is bigger:

```
x1 = linspace(-1,1, 301);   % note 301
[x,y] = meshgrid(x1,x1);
h = pcolor(x,y,sin(x)+y);
```

(Default Octave behaviour is the figure size is not dependent of its resolution.)

I'm sure there are more important things to fix about 3D plots, but here's one more ;-)

Great project so far!

**Comment by blink1073** [comment]
Thanks for the report, @cbm755, getting plotting right has proven to be tricky ;-)

**Comment by thomasjm** [comment]
@blink1073 I have a similar issue that I think is related: in Octave itself, the code

```
tx = ty = linspace (-8, 8, 41)';
[xx, yy] = meshgrid (tx, ty);
r = sqrt (xx .^ 2 + yy .^ 2) + eps;
tz = sin (r) ./ r;
mesh (tx, ty, tz);
```

produces this nice-looking plot:

![image](https://cloud.githubusercontent.com/assets/1634990/20551614/026c8986-b0f7-11e6-9f31-36615ba76c30.png)

but in a Jupyter notebook, it produces this:

![image](https://cloud.githubusercontent.com/assets/1634990/20551628/2515f08a-b0f7-11e6-8ef3-7ca1751e2186.png)

Any idea what's going on/how this could be improved? Thanks!

**Comment by blink1073** [comment]
Hi @thomasjm, I am seeing the desired plot from a notebook with an Octave kernel.    What operating system are you using?

**Comment by thomasjm** [comment]
I'm on Ubuntu 16.04. I was guessing it has something to do with a different plotting backend being used (gnuplot in the notebook, vs. something else in the Octave UI?)

**Comment by blink1073** [comment]
@thomasjm, can you please try it as the following, it could be that the png output is not working properly with your version of gnuplot.

```
%plot -f svg
tx = ty = linspace (-8, 8, 41)';
[xx, yy] = meshgrid (tx, ty);
r = sqrt (xx .^ 2 + yy .^ 2) + eps;
tz = sin (r) ./ r;
mesh (tx, ty, tz);
```

**Comment by thomasjm** [comment]
Aha, that worked! 

I've been using gnuplot 5.0.1. I just tried upgrading to 5.0.5, which seems to be the latest, but the same problem still happens. What version of gnuplot are you using?

**Comment by blink1073** [comment]
I have had problems with some png saving on OSX as well, using gnuplot 0.5.3.  I believe it is related to this bug: http://savannah.gnu.org/bugs/?44458.

**Comment by thomasjm** [comment]
Hmm, I've been digging into this a little more. The code that the kernel uses looks like 

```
_figHandles = get(0, 'children');
for _fig=1:length(_figHandles),
    ...
    try,
       ...
       imwrite(uint8(_image*255), _filename);
    catch,
       print(_handle, _filename, '-r%(res)s');
    end,
    close(_handle);
end;
```

It seems that it prefers to use `imwrite`, and will only use `print` if there's an exception.

When I type these commands into my normal Octave GUI, I notice that `imwrite(uint8(_image*255), "test.png");` produces the ugly image, but `print -dpng test.png` produces the nice image. Is it possible that `imwrite` just isn't the right thing to use in the case of PNGs?

**In fact**, I notice that `imwrite(uint8(_image*255), "test.svg");` actually throws an exception:

```
error: Magick++ exception: Magick: No encode delegate for this image format (test.svg) reported by magick/constitute.c:2261 (WriteImage)
```

so probably the only reason `%plot -f svg` works is because it's falling back on `print`, and `print` is all-around better to use. What do you think?

**Comment by blink1073** [comment]
Ah, I think the problem is that we haven't properly scaled the image to 8 bit.  I still think `imwrite` should be the preferred option, since it gives you an actual image instead of a "figure", but I think we can overcome the scaling problem.

**Comment by thomasjm** [comment]
As far as I can tell, `imwrite` is used to more or less just write matrices into an image file. The `_image` that we end up writing in my case is a matrix of doubles of size 41x41. Even if we scale it up, we'll never get anything but a flat grayscale picture, not the nice "3D" view that we see in the nicer image. I'm not sure how Octave produces the colors and the 3D perspective, but it seems like it might be producing postscript commands and invoking a PS renderer. In any case, `imwrite` seems like the wrong thing (for 3D plots anyway).

Also, I'm pretty sure `imwrite` will never be the right thing for formats like SVG, which are not "matrix-like" images.

Googling around, I see people are recommending the use of `print` for exporting figures: http://stackoverflow.com/questions/17211048/matlab-imwrite-quality

**Comment by blink1073** [comment]
Yeah, I think we can fix this issue without using `imwrite` at all, but still need to make sure the original use case for #40 is still met.  Handling plotting "automatically" is a real PITA.

**Comment by blink1073** [comment]
This is a better check to see if it is an image: `strcmp(get(get(get(gcf, 'children'), 'children'), 'type'), 'image')`.  That returns 0 for a `surf` and 1 for an `imshow`.

**Comment by blink1073** [comment]
This is fixed not thanks to #51, thanks again @thomasjm!
